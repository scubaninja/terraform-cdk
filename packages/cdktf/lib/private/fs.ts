import { execSync } from "child_process";
import * as fs from "fs";
import * as path from "path";
import * as os from "os";

// Full implementation at https://github.com/jprichardson/node-fs-extra/blob/master/lib/copy-sync/copy-sync.js
export function copySync(src: string, dest: string) {
  function copyItem(p: string) {
    const sourcePath = path.resolve(src, p);
    const stat = fs.statSync(sourcePath);
    if (stat.isFile()) {
      fs.copyFileSync(sourcePath, path.resolve(dest, p));
    }
    if (stat.isDirectory()) {
      walkSubfolder(p);
    }
  }

  function walkSubfolder(p: string) {
    const sourceDir = path.resolve(src, p);
    fs.mkdirSync(path.resolve(dest, p), { recursive: true });
    fs.readdirSync(sourceDir).forEach((item) => copyItem(path.join(p, item)));
  }

  walkSubfolder(".");
}

export function archiveSync(src: string, dest: string) {
  try {
    if (os.platform() === "win32") {
      execSync("where zip");
    } else {
      execSync("which zip");
    }
  } catch {
    console.error(`Unable to find "zip".`);
    process.exit(1);
  }

  execSync(`zip -r ${dest} ${src}`);
}
