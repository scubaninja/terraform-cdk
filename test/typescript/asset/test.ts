/**
 * Testing synthing typescript to json
 *
 * @group typescript
 * @group asset
 */
import { TestDriver } from "../../test-helper";
import * as path from "path";
import * as fs from "fs";

describe("full integration test", () => {
  let driver: TestDriver;

  beforeAll(async () => {
    driver = new TestDriver(__dirname);
    await driver.setupTypescriptProject();
    driver.copyFiles("local-asset.txt");
    driver.copyFolders("fixtures");
  });

  test("synth", async () => {
    await driver.synth();
    expect(driver.synthesizedStack("stack")).toMatchSnapshot();
  });

  test("file asset copied", async () => {
    await driver.synth();
    expect(
      fs.readFileSync(
        path.resolve(
          driver.stackDirectory("stack"),
          "localasset/local-asset.txt"
        ),
        "utf-8"
      )
    ).toMatchSnapshot();
  });

  test("folder asset copied", async () => {
    await driver.synth();
    expect(
      fs.readFileSync(
        path.resolve(driver.stackDirectory("stack"), "fixtures/a.txt"),
        "utf-8"
      )
    ).toMatchSnapshot();
    expect(
      fs.readFileSync(
        path.resolve(driver.stackDirectory("stack"), "fixtures/b.txt"),
        "utf-8"
      )
    ).toMatchSnapshot();
    expect(
      fs.readFileSync(
        path.resolve(driver.stackDirectory("stack"), "fixtures/foo/bar/c.txt"),
        "utf-8"
      )
    ).toMatchSnapshot();
  });

  test("zip file created", async () => {
    await driver.synth();

    const stat = fs.statSync(
      path.resolve(driver.stackDirectory("stack"), "zippedfixtures/archive.zip")
    );
    expect(stat.isFile()).toBe(true);
    expect(stat.size).toBe(1799);
  });
});
