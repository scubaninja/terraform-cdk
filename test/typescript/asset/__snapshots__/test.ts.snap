// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`full integration test file asset copied 1`] = `"Hello Asset World"`;

exports[`full integration test folder asset copied 1`] = `"This is a"`;

exports[`full integration test folder asset copied 2`] = `"This is b"`;

exports[`full integration test folder asset copied 3`] = `"this is c"`;

exports[`full integration test synth 1`] = `
"{
  \\"//\\": {
    \\"metadata\\": {
      \\"version\\": \\"stubbed\\",
      \\"stackName\\": \\"stack\\"
    }
  },
  \\"assets\\": {
    \\"localasset\\": {
      \\"path\\": \\"../../../local-asset.txt\\"
    },
    \\"fixtures\\": {
      \\"path\\": \\"../../../fixtures\\"
    },
    \\"zippedfixtures\\": {
      \\"path\\": \\"../../../fixtures\\"
    }
  }
}"
`;
